docs/design/cronjob.md

```diff
@@ -0,0 +1,118 @@
+useful for generating data for test or research. 
+
+## Function Specification
+
+The kubernetes [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) would be a best example when considering the functionality or behaviour of our own batch `CronJob`,
+therefore, our API and controller would be much similar to the original `CronJob` in kubernetes except the field that we focus.
+
+### API
+
+```go
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type CronJob struct {
+	metav1.TypeMeta `json:",inline"`
+
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired behavior of a CronJob
+	Spec CronJobSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Current status of CronJob
+	// +optional
+	Status CronJobStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+type CronJobSpec struct {
+	// The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
+    Schedule string `json:"schedule" protobuf:"bytes,1,opt,name=schedule"`
+	// Template is a template based on which the Job should run.
+
+	Template JobSpec `json:"template" protobuf:"bytes,2,opt,name=spec"`
+    // Optional deadline in seconds for starting the job if it misses scheduled
+    // time for any reason.  Missed jobs executions will be counted as failed ones.
+    // +optional
+    StartingDeadlineSeconds *int64 `json:"startingDeadlineSeconds,omitempty" protobuf:"varint,3,opt,name=startingDeadlineSeconds"`
+
+    // Specifies how to treat concurrent executions of a Job.
+    // Valid values are:
+    // - "Allow" (default): allows CronJobs to run concurrently;
+    // - "Forbid": forbids concurrent runs, skipping next run if previous run hasn't finished yet;
+    // - "Replace": cancels currently running job and replaces it with a new one
+    // +optional
+    ConcurrencyPolicy ConcurrencyPolicy `json:"concurrencyPolicy,omitempty" protobuf:"bytes,4,opt,name=concurrencyPolicy,casttype=ConcurrencyPolicy"`
+
+    // This flag tells the controller to suspend subsequent executions, it does
+    // not apply to already started executions.  Defaults to false.
+    // +optional
+    Suspend *bool `json:"suspend,omitempty" protobuf:"varint,5,opt,name=suspend"`
+    
+    // The number of successful finished jobs to retain.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    // Defaults to 3.
+    // +optional
+    SuccessfulJobsHistoryLimit *int32 `json:"successfulJobsHistoryLimit,omitempty" protobuf:"varint,6,opt,name=successfulJobsHistoryLimit"`
+
+    // The number of failed finished jobs to retain.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    // Defaults to 1.
+    // +optional
+    FailedJobsHistoryLimit *int32 `json:"failedJobsHistoryLimit,omitempty" protobuf:"varint,7,opt,name=failedJobsHistoryLimit"`
```

Comment:
Did not check code in upstream, but how do we trace this history?

Reply:
By browsing all the jobs in the cluster and deleting some of the jobs according to this attribute.